input/full.go

```diff
@@ -137,6 +142,12 @@ func CollectFull(ctx context.Context, server *state.Server, connection *sql.DB,
 		return
 	}
 
+	// CollectAllSchemas relies on GetBufferCache to access the filenode OIDs before that data is discarded
+	select {
+	case <-ctx.Done():
+	case ts.BufferCache = <-bufferCacheReady:
+	}
```

Comment:
Since querying `pg_buffercache` can be slow (~12 seconds for 200 GB of shared buffers), the query is performed in a separate goroutine so full snapshot collection can proceed without blocking on the buffer cache data, until it's needed.

Reply:
Yes, this will run an additional database connection concurrently when generating a full snapshot, so we could run into the connection limit. But I don't think that's likely to happen because by default we have a limit of 10 connections per server.

Reply:
> But I don't think that's likely to happen because by default we have a limit of 10 connections per server.

Ah true, it's per server, not per collector - yeah in that case, it's at most one additional connection per server, so it's gonna be okay 👍 

I had another question around here: you say "until it's needed", and it's true - if the full snapshot finishes running (outputting buffer cache) before this `ts.BufferCache = <-bufferCacheReady`, it'll output the data of 10 mins ago, correct? IOW, full snapshot with a timely buffer cache info can be obtained only when this goroutine finishes before the "outputting buffer cache" part is done.
I think it's better than making a full snapshot runs longer (so I'm not opposed to it), but just wanted to double check if my understanding is correct.

Reply:
> if the full snapshot finishes running (outputting buffer cache) before this `ts.BufferCache = <-bufferCacheReady`, it'll output the data of 10 mins ago, correct?

No, the code is blocking on the message from that channel. It will wait for the new buffer cache stats to be available. For our production server that has 198 GB of shared buffers, querying `pg_buffercache` takes 12 seconds. I think that's short enough to not be an issue.